lib/firmware/fsm_msg_nano.h

```diff
@@ -0,0 +1,383 @@
+
+    // Determine what type of prompt to display
+    bool needs_confirm = true;
+    bool is_transfer = false;
+    if (strlen(representative_address) > 0) {
+        // always confirm representative change
+    } else if (!is_send) {
+        // don't bother confirming pure receives
+        needs_confirm = false;
+    } else if (msg->has_address_type &&
+               msg->address_type == OutputAddressType_TRANSFER &&
+               !nano_path_mismatched(coin, msg->link_recipient_n,
+                                     msg->link_recipient_n_count)) {
+        is_transfer = true;
+    }
+
+    if (needs_confirm) {
+        if (strlen(representative_address) > 0) {
+            if (!confirm(ButtonRequestType_ButtonRequest_ConfirmOutput,
+                         "Representative", "Set account representative to %s?",
+                         representative_address)) {
+                fsm_sendFailure(FailureType_Failure_ActionCancelled, "Signing cancelled");
+                layoutHome();
+                return;
+            }
+        }
+
+        char amount_string[60];
+        memset(amount_string, 0, sizeof(amount_string));
+        bn_format(&balance_delta,
+                  NULL, NULL,
+                  coin->decimals, 0, false,
+                  amount_string, sizeof(amount_string));
+
+        if (is_transfer) {
+            // Confirm transfer between own accounts
+            char account_str[NODE_STRING_LENGTH];
+            if (!nano_bip32_to_string(account_str, sizeof(account_str), coin, msg->link_recipient_n,
+                                      msg->link_recipient_n_count) &&
+                !bip32_path_to_string(account_str, sizeof(account_str),
+                                      msg->link_recipient_n, msg->link_recipient_n_count)) {
+                strlcpy(account_str, recipient_address, sizeof(account_str));
+            }
+
+            if (!confirm(ButtonRequestType_ButtonRequest_SignTx,
+                         "Transfer", "Send %s %s to %s?",
+                         amount_string,
+                         coin->coin_shortcut,
+                         account_str)) {
+                fsm_sendFailure(FailureType_Failure_ActionCancelled, "Signing cancelled");
+                layoutHome();
+                return;
+            }
+        } else if (is_send) {
+            // Regular transfer
+            if (!confirm(ButtonRequestType_ButtonRequest_SignTx,
+                         "Send", "Send %s %s to %s?",
+                         amount_string,
+                         coin->coin_shortcut,
+                         recipient_address)) {
+                fsm_sendFailure(FailureType_Failure_ActionCancelled, "Signing cancelled");
+                layoutHome();
+                return;
+            }
+        }
+    }
+
+    // Sign hash and return the signature
+    uint8_t signature[64];
+    memset(signature, 0, sizeof(signature));
+    hdnode_sign_digest(node, block_hash, signature, NULL, NULL);
```

Comment:
`fsm_getDerivedNode()` has the node as a static inside it, which would get overwritten by the call here: https://github.com/keepkey/keepkey-firmware/pull/82/files#diff-a51ea15f50216fc3c1b6614a895b8027R223

Reply:
Nice catch! :+1:

Reply:
Usual fix is to copy the node and save it for later. memzero under all the failure condition branches.

Reply:
Since there seems to be some caching involved in that function, I was thinking of just invalidating the pointer and then calling the `fsm_getDerivedNode` before signing if `node == NULL`.

Reply:
Or just wait to calculate it until it's needed.